blockmap.js

```diff
@@ -0,0 +1,126 @@
+'use strict';
+
+const BlockMap = require('blockmap');
+const crypto = require('crypto');
+
+function getNotDiscardedChunks(disk, blockSize, capacity) {
+	const chunks = [];
+	const discardedChunks = disk.getDiscardedChunks();
+	let lastStart = 0;
+	for (let discardedChunk of discardedChunks) {
+		chunks.push([lastStart, discardedChunk.start - 1]);
+		lastStart = discardedChunk.end + 1;
+	}
+	if (lastStart < capacity) {
+		chunks.push([lastStart, capacity - 1]);
+	}
+	return chunks;
+}
+
+function mergeBlocks(blocks) {
+	// Merges adjacent blocks in place (helper for getBlockMap).
+	if (blocks.length > 1) {
+		let last = blocks[0];
+		for (let i = 1; i < blocks.length; i++) {
+			let block = blocks[i];
+			if ((block[0] >= last[0]) && (block[0] <= last[1] + 1)) {
+				last[1] = block[1];
+				blocks.splice(i, 1);
+				i--;
+			} else {
+				last = block;
+			}
+		}
+	}
+}
+
+function streamSha256(stream, callback) {
+	const hash = crypto.createHash('sha256');
+	hash.setEncoding('hex');
+	stream.pipe(hash);
+	stream.on('error', callback);
+	stream.on('end', function() {
+		hash.end();
+		callback(null, hash.read());
+	});
+}
+
+function getRanges(disk, blocks, blockSize, callback) {
+	const result = new Array(blocks.length);
+	let wait = blocks.length;
+	let block, i, start, length;
+	let error = false;
+	for (i = 0; i < blocks.length; i++) {
+		block = blocks[i];
+		start  = block[0] * blockSize;
+		length = (block[1] - block[0] + 1) * blockSize;
+		(function(i, start, length, block) {
+			disk.getStream(start, length, function(err, stream) {
+				if (err) {
+					if (!error) {
+						error = true;
+						callback(err);
+					}
+					wait -= 1;
```

Comment:
Why do you implement this ad-hoc sync primitive here instead of using https://github.com/caolan/async or Promises? This is brittle if you forget to decrease the counter or forget to return and double decrease

Reply:
Replaced with promises